pkg/controller/controller_test.go

```diff
@@ -66,3 +72,92 @@ func setupData(ctx context.Context, t *testing.T, trs []*v1beta1.TaskRun) inform
 	c.ClearActions()
 	return tri
 }
+
+func TestReconciler_handleTaskRun(t *testing.T) {
+
+	tests := []struct {
+		name       string
+		tr         *v1beta1.TaskRun
+		shouldSign bool
+	}{
+		{
+			name: "complete, already signed",
+			tr: &v1beta1.TaskRun{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{signing.ChainsAnnotation: "true"},
+				},
+				Status: v1beta1.TaskRunStatus{
+					Status: duckv1beta1.Status{
+						Conditions: []apis.Condition{{Type: apis.ConditionSucceeded}},
+					}},
+			},
+			shouldSign: false,
+		},
+		{
+			name: "complete, not already signed",
+			tr: &v1beta1.TaskRun{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{},
+				},
+				Status: v1beta1.TaskRunStatus{
+					Status: duckv1beta1.Status{
+						Conditions: []apis.Condition{{Type: apis.ConditionSucceeded}},
+					}},
+			},
+			shouldSign: true,
+		},
+		{
+			name: "not complete, not already signed",
+			tr: &v1beta1.TaskRun{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{},
+				},
+				Status: v1beta1.TaskRunStatus{
+					Status: duckv1beta1.Status{
+						Conditions: []apis.Condition{},
+					}},
+			},
+			shouldSign: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Mock out SignTaskRun
+			m, cleanup := setupMocks()
+			defer cleanup()
+			ctx, _ := rtesting.SetupFakeContext(t)
+			c := fakepipelineclient.Get(ctx)
+
+			r := &Reconciler{
+				Base: &reconciler.Base{
+					PipelineClientSet: c,
+				},
+				Logger: logtesting.TestLogger(t),
+			}
+			if err := r.handleTaskRun(ctx, tt.tr); err != nil {
+				t.Errorf("Reconciler.handleTaskRun() error = %v", err)
+			}
+			if m.signed != tt.shouldSign {
```

Comment:
Any reason not to use `IsSigned` here?

Reply:
I mocked out the function that actually does the signing (it would require a fake API server and more boilerplate.

Reply:
I was thinking `IsSigned` because it just takes a `TaskRun`. I think this makes more sense in an e2e test, but since we're faking the signing anyway, it just seemed like `IsSigned` could be used here to check the annotations that are already configured in this test's setup, or have the mock signing function add the annotation itself.